pbs/usersync.go

```diff
@@ -10,33 +10,86 @@ import (
 	"strings"
 	"time"
 
+	"errors"
 	"github.com/golang/glog"
 	"github.com/julienschmidt/httprouter"
 	"github.com/prebid/prebid-server/ssl"
-	metrics "github.com/rcrowley/go-metrics"
+	"github.com/rcrowley/go-metrics"
 )
 
-var cookie_domain string
-var external_url string
-var recaptcha_secret string
+// Recaptcha code from https://github.com/haisum/recaptcha/blob/master/recaptcha.go
+const RECAPTCHA_URL = "https://www.google.com/recaptcha/api/siteverify"
+const COOKIE_NAME = "uids"
 
-type PBSCookie struct {
-	UIDs     map[string]string `json:"uids,omitempty"`
-	OptOut   bool              `json:"optout,omitempty"`
-	Birthday *time.Time        `json:"bday,omitempty"`
+const (
+	USERSYNC_OPT_OUT     = "usersync.opt_outs"
+	USERSYNC_BAD_REQUEST = "usersync.bad_requests"
+	USERSYNC_SUCCESS     = "usersync.%s.sets"
+)
+
+type UserSyncDeps struct {
+	Cookie_domain    string
+	External_url     string
+	Recaptcha_secret string
+	Metrics          metrics.Registry
+}
+
+// UserSyncMap is cookie which stores the user sync info for all of our bidders.
+//
+// To get an instance of this from a request, use ParseUserSyncMapFromRequest.
+// To write an instance onto a response, use SetCookieOnResponse.
+type UserSyncMap interface {
```

Comment:
Implemented this in #106.

The two look pretty similar in complexity to me... but I think I give a slight edge to the concrete struct. This is how I'm thinking about it:

Concrete struct benefits:
- nil-safety is possible (though it requires some extra branchy code).
- `nil` is much clearer to mobile apps, since there's no such thing as a UserSync there.

Interface benefits:
- No custom JSON code, so there's less room for bugs

In this case I think the confusion caused by a UserSync map on mobile requests is probably the most important part.... so gonna go with that one.